### PR TITLE
trace|cpu & gpu implementation

### PR DESCRIPTION
**Pull request by bosr**
As per https://github.com/OpenMined/PySyft/issues/611

- cpu & gpu implementation
- simple unit tests are written here
- for gpu testing, I did the following from Syft

```
a = sc.FloatTensor(np.array([[-1, 2], [6, 7.3]]).astype('float'))
a.trace()  # returns 6.3
a = a.gpu()
a.trace()  # the same

# testcase with larger size than the groupsize
b = sc.FloatTensor(np.ones((96, 96)).astype('float'))
b.trace()  # returns 96
b = b.gpu()
b.trace()  # the same
```

- [x] an integration test in PySyft demonstrating the correct CPU and GPU operation implemented over a FloatTensor while connected to a Unity backend
- [x] a Unit Test in OpenMined/OpenMined demonstrating the correct operation on a FloatTensor
- [x] inline documentation in the python code. For inspiration on inline documentation, please check out -PyTorch's documentation for this operator.
- [x] Link your Pull Request back to this Issue so that it gets closed appropriately when the PR is merged.

### CODE REVIEW

**Comment by diegoalejogm** [comment]
Well done!! Let’s wait for the tests to end running. We’re starting out with unit tests if you wanna check out #144

**Comment by bosr** [comment]
Cool, I'll check this!

**Comment by bosr** [comment]
wip wip ;)
- rebased (a bit painful because of indentation causing conflicts. Anyway)
- I'm refactoring the gpu trace to not rely on cpu sum at the end (it's not that much, but I'm learning 😆 )
- I'll add integration test as @diegoalejogm showed (very cool)

**Comment by adamrainsby** [comment]
> a bit painful because of indentation causing conflicts

Hopefully not for long #157

**Comment by diegoalejogm** [comment]
@bosr sorry to tell you that there are some more merge conflicts. There's been many refactoring tasks this week. Hopefully it is stable now 👍

**Comment by bosr** [comment]
Yep no problem, I'm keeping up on slack. the codebase is constantly improving, that's cool!!  

Cheers

**Comment by bosr** [comment]
Several changes:
- O(n) threads for O(n^2) matrices
- gpu impl uses 2 kernels: DiagonalKernel and a separate more generic kernel `Reduce1DSumKernel` for the parallel reduction using shared mem and no more threads than needed
- integration tests for both cpu and gpu
- trace returns a float, not a FloatTensor anymore (so this is not currently usable from Syft, see below).

I suggest returning a float and not a tensor because I think we need an equivalent to Cmd message for the return side (Unity -> Syft). I'll write an issue very soon (tomorrow morning).

**Comment by diegoalejogm** [comment]
Are you ready to merge? Let me know by removing the flag

File: UnityProject/Assets/OpenMined/Syft/Math/Shaders/FloatTensorShaders.compute [hunk 1]
```diff
@@ -540,6 +555,34 @@ void Tanh (uint3 id: SV_DispatchThreadID) {
...
 }
 
+#pragma kernel Reduce1DSum
+#define REDUCE1D_NUMTHREADS 8
+
+RWStructuredBuffer<int> Reduce1DSumNumgroups;
+RWStructuredBuffer<float> Reduce1DSumData;
+RWStructuredBuffer<float> Reduce1DSumResult;
+
+groupshared float sharedMem[REDUCE1D_NUMTHREADS * 1 * 1];
```
Comment (diegoalejogm) [inline UnityProject/Assets/OpenMined/Syft/Math/Shaders/FloatTensorShaders.compute]: This is pretty awesome! Didn't know one could do this in HLSL

**Comment by diegoalejogm** [comment]
That's a pretty cool implementation, congrats!

I am no expert in the field, but found this very comprehensive resource on how optimize parallel reduction in GPUs: http://developer.download.nvidia.com/compute/cuda/1.1-Beta/x86_website/projects/reduction/doc/reduction.pdf

I think we might still be able to improve upon your current proposal. Nevertheless I think it's not easy to come up with this, so again, congrats @bosr. I am very proud.

**Comment by bosr** [comment]
Thanks 😊  I'm learning HLSL actually, so you may know more than me!

Regarding ReduceSum, I anticipate that leveraging a common kernel will ease per-platform tuning, and other few kernels that are actually reductions. At least for trace it helped me untangle things.

Actually, in my current implementation, there may be bank conflicts, but
- maybe not on CUDA: we process `float`s which each one occupies a sharedmem bank
- this kernel is the one place where to improve this
- we do not have gpu profiling yet (I'll create an issue for this)

Thanks for the link, cool!! I have some links too and will add those to gpu docs.

File: UnityProject/Assets/OpenMined/Syft/Math/Shaders/FloatTensorShaders.compute [hunk 2]
```diff
@@ -540,6 +555,34 @@ void Tanh (uint3 id: SV_DispatchThreadID) {
 	TanhResult[id.x] = tanh(TanhData[id.x]);
 }
 
+#pragma kernel Reduce1DSum
+#define REDUCE1D_NUMTHREADS 8
+
+RWStructuredBuffer<int> Reduce1DSumNumgroups;
```
[the remark this answers is not in the document]
Comment (bosr) [reply UnityProject/Assets/OpenMined/Syft/Math/Shaders/FloatTensorShaders.compute]: yup, will fix this today, good catch ;)

**Comment by bosr** [comment]
@diegoalejogm rebased and ready to merge (I don't think I can remove flags)

**Comment by bosr** [comment]
"we do not have gpu profiling yet", stupid of me: https://docs.unity3d.com/Manual/ProfilerGPU.html

**Comment by bosr** [comment]
Great pdf, thank you Diego!! Some very neat ideas to implement!

**Pull request merged**